### ISSUE TITLE
then_signal_semaphore introduces VUID-vkQueuePresentKHR-pWaitSemaphores-03268

### ISSUE THREAD

**Issue by LeonMatthes** [vulkano-rs/vulkano#1885]
# Template

*   Version of vulkano: "0.29"
*   OS: Fedora 35
*   GPU (the selected PhysicalDevice): GTX 1070 (desktop)
*   GPU Driver: nvidia
*   Upload of a reasonably minimal complete `main.rs` file that demonstrates the issue: https://github.com/LeonMatthes/Vulkano-Hello-Triangle/commit/72720e934502dc967cc75694d326a7dd9277cb5a

# Issue

I've tried following vulkan-tutorial.com and translating the code to vulkano as closely as possible.
The code that errors should therefore be the minimal required to get to the triangle.
The problem occurs when I run the program in debug mode (with validation layers enabled).

At first, the triangle will render correctly, however, once the window is resized, the validation layers issue 2 instances of `VUID-vkQueuePresentKHR-pWaitSemaphores-03268` and the application becomes unresponsive.

```
VUID-vkQueuePresentKHR-pWaitSemaphores-03268(ERROR / SPEC): msgNum: 622825338 - Validation Error: [ VUID-vkQueuePresentKHR-pWaitSemaphores-03268 ] Object 0: handle = 0x55d11b9896b0, type = VK_OBJECT_TYPE_QUEUE; Object 1: handle = 0x980f360000000011, type = VK_OBJECT_TYPE_SEMAPHORE; | MessageID = 0x251f8f7a | vkQueuePresentKHR: Queue VkQueue 0x55d11b9896b0[] is waiting on pWaitSemaphores[0] (VkSemaphore 0x980f360000000011[]) that has no way to be signaled. The Vulkan spec states: All elements of the pWaitSemaphores member of pPresentInfo must reference a semaphore signal operation that has been submitted for execution and any semaphore signal operations on which it depends (if any) must have also been submitted for execution (https://vulkan.lunarg.com/doc/view/1.3.211.0/linux/1.3-extensions/vkspec.html#VUID-vkQueuePresentKHR-pWaitSemaphores-03268)
    Objects: 2
        [0] 0x55d11b9896b0, type: 4, name: NULL
        [1] 0x980f360000000011, type: 5, name: NULL
VUID-vkQueuePresentKHR-pWaitSemaphores-03268(ERROR / SPEC): msgNum: 622825338 - Validation Error: [ VUID-vkQueuePresentKHR-pWaitSemaphores-03268 ] Object 0: handle = 0x55d11b9896b0, type = VK_OBJECT_TYPE_QUEUE; Object 1: handle = 0x980f360000000011, type = VK_OBJECT_TYPE_SEMAPHORE; | MessageID = 0x251f8f7a | vkQueuePresentKHR: Queue VkQueue 0x55d11b9896b0[] is waiting on pWaitSemaphores[0] (VkSemaphore 0x980f360000000011[]) that has no way to be signaled. The Vulkan spec states: All elements of the pWaitSemaphores member of pPresentInfo must reference a semaphore signal operation that has been submitted for execution and any semaphore signal operations on which it depends (if any) must have also been submitted for execution (https://vulkan.lunarg.com/doc/view/1.3.211.0/linux/1.3-extensions/vkspec.html#VUID-vkQueuePresentKHR-pWaitSemaphores-03268)
    Objects: 2
        [0] 0x55d11b9896b0, type: 4, name: NULL
        [1] 0x980f360000000011, type: 5, name: NULL
```

I know that the swapchain should be invalidated once the resize event happens. I'm checking the appropriate return values of acquire_next_image and aborting the render operation in that case.
There are however multiple unexpected behaviors here:

1. The `ResizeEvent` is issued by winit, but the next acquire_next_image function does not fail, and still returns an `Ok` result and suboptimal == false.
2. Removing the line 475 (`.then_signal_semaphore()`) seems to fix the issue

I believe what is happening here is that the application hangs on the `future.wait(None)` call as the semaphore created by vulkano is never signalled.
I would just remove the `then_signal_semaphore()` code, however the `then_execute` and `then_swapchain_present` calls happen on two queues that *might* be two separate Queues - even though they are not in practice on my GTX 1070.
As the [sync docs](https://docs.rs/vulkano/latest/vulkano/sync/index.html#between-two-different-gpu-queues) state, if these two queues are ever actually different, not using the semaphore would always result in a runtime error.

## Expected behavior
Either:
- Not get the VUID-vkQueuePresentKHR-pWaitSemaphores-03268 error when using the result of swapchain::acquire_next_image after a resize event.
- Not get a valid result from the swapchain::acquire_next_image if the window has already been resized.

It is also possible I'm calling vulkano incorrectly somewhere, but as the docs state that vulkano aims to always detect incorrect API usage, this might still indicate a bug.

Thanks for your help.

**Comment by LeonMatthes** [comment]
This might somehow be related to: #1705

**Comment by Rua** [comment]
Your immediate problem can be solved by removing the call to `then_signal_semaphore`. `then_swapchain_present` will work fine without it, as seen in Vulkano's own examples.

The problem you report is valid, though, and is a long-standing problem with Vulkano's synchronisation. Some of the problems within command buffers are dealt with in the upcoming release, but queue synchronisation, the use of semaphores and fences, the lack of timeline semaphores and events, are known problems and quite difficult to solve correctly. They definitely have my attention, but it will probably have to be fixed in small bits at a time in order to not introduce too many new bugs.

**Comment by LeonMatthes** [comment]
Hm, in the meantime I've done some more investigating.
If the `graphics_q` and `surface_q` in my example are indeed separate queues instances, I think then I'd actually need the semaphore, otherwise I'd run into a problem.

For further debugging of this, I've dug into the API dump and Semaphore/GPUFuture code a bit.
The API dump can be found in the attached [freeze.log](https://github.com/vulkano-rs/vulkano/files/8597350/freeze.log) file.

Interestingly, the API Dump indicates that Vulkano actually issues 3 vkQueuePresentKHR calls that all wait on the same semaphore, followed by a vkQueueWaitIdle.
I'd suspect the issue here is that only one of the Present calls can correctly wait on the Semaphore and the other ones will never be signaled.
This is also what then causes the freeze when vkQueueWaitIdle is called, as the queue is never emptied.

This issue actually seems to be caused by a bug in the PresentFuture's code.
When a PresentFuture's `flush` fails with "Err(FlushError::OutOfDate)", the Drop code of the PresentFuture thinks that the future has not yet been correctly flushed and does a second flush, which causes the second vkQueuePresentKHR.

I was able to get around this issue with this code:
``` rust
        let draw_and_present = acquire_future
            .then_execute(self.graphics_q.clone(), command_buffer.clone())
            .unwrap()
            .then_signal_semaphore()
            .then_swapchain_present(
                self.surface_q.clone(),
                self.swapchain.swapchain.clone(),
                image_i,
            );
        match draw_and_present.flush() {
            Err(FlushError::OutOfDate) => {
                println!("Out of date!");
                unsafe {
                    draw_and_present.signal_finished();
                }
                return;
            }
            Err(e) => {
                println!("{:?}", e);
            }
            Ok(()) => (),
        };

        match draw_and_present.then_signal_fence_and_flush() {
            Ok(future) => {
                future.wait(None).unwrap();
            }
            Err(FlushError::OutOfDate) => {
                println!("Fence - Out of date!");
            }
            Err(e) => {
                println!("Fence - Failed to flush future: {:?}", e);
            }
        }

```
The issue here is that I need to call "signal_finished" in the Err handler of the draw_and_present future because of the error in the PresentFuture mentioned earlier.

In my opinion this is still the better way to handle the flushing, as every flush represents one vkQueue... command, which should have its own error handling.

It still also has issue #1705 that the fence is created by submitting an empty vkQueuePresent.

Furthermore, if I ever do a resize, I'll get this:
```
VUID-vkDestroySemaphore-semaphore-01137(ERROR / SPEC): msgNum: -1588160456 - Validation Error: [ VUID-vkDestroySemaphore-semaphore-01137 ] Object 0: handle = 0x56273f5afd00, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xa1569838 | Cannot call vkDestroySemaphore on VkSemaphore 0x9fde6b0000000014[] that is currently in use by a command buffer. The Vulkan spec states: All submitted batches that refer to semaphore must have completed execution (https://vulkan.lunarg.com/doc/view/1.3.211.0/linux/1.3-extensions/vkspec.html#VUID-vkDestroySemaphore-semaphore-01137)
    Objects: 1
        [0] 0x56273f5afd00, type: 3, name: NULL
VUID-vkDestroySemaphore-semaphore-01137(ERROR / SPEC): msgNum: -1588160456 - Validation Error: [ VUID-vkDestroySemaphore-semaphore-01137 ] Object 0: handle = 0x56273f5afd00, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xa1569838 | Cannot call vkDestroySemaphore on VkSemaphore 0x980f360000000011[] that is currently in use by a command buffer. The Vulkan spec states: All submitted batches that refer to semaphore must have completed execution (https://vulkan.lunarg.com/doc/view/1.3.211.0/linux/1.3-extensions/vkspec.html#VUID-vkDestroySemaphore-semaphore-01137)
    Objects: 1
        [0] 0x56273f5afd00, type: 3, name: NULL
```
Which is probably because then nothing ever waits on the semaphore signalled by the command_buffer, but I'm not entirely sure.

So as an immediate action I'd say the error handling when flushing a PresentFuture should be improved, so that then dropping the semaphore doesn't submit a duplicate flush.

Maybe changing the current GpuFuture a bit to represent an acyclic graph instead of the current semi-linear program flow would make sense.
i.e.: when calling `then_signal_semaphore`, actually return **two** objects - one SemaphoreFuture and the previous CommandBufferExecFuture.
Same with a `then_signal_fence` call.
That would make it possible to specify which semaphores/fences to use for what.

**Comment by Rua** [comment]
If you feel up to it, can you make the changes to `then_signal_semaphore` and create a new PR? I'm currently busy with other Vulkano work, and any help would be greatly appreciated.

**Comment by LeonMatthes** [comment]
@Rua I think I was able to fix this in #1889
Please make sure to review the PR thoroughly though. I'm still at the "Hello Triangle" stage of learning Vulkan, so may have missed some Vulkan details regarding the queue wait, etc.

**Comment by LeonMatthes** [comment]
Thanks for merging my PR. That solves the issue in my book :)
